libbpf-tools/syscount.bpf.c

```diff
@@ -0,0 +1,103 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020 Anton Protopopov
+//
+// Based on syscount(8) from BCC by Sasha Goldshtein
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+#include "syscount.h"
+
+const volatile bool count_by_process = false;
+const volatile bool measure_latency = false;
+const volatile bool filter_failed = false;
+const volatile int filter_errno = false;
+const volatile pid_t filter_pid = 0;
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u64);
+	__type(value, u64);
+} start SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u32);
+	__type(value, struct data_t);
+} data SEC(".maps");
+
+static __always_inline
+void *bpf_map_lookup_or_try_init(void *map, void *key, void *init)
+{
+	void *val;
+	int err;
+
+	val = bpf_map_lookup_elem(map, key);
+	if (val)
+		return val;
+
+	err = bpf_map_update_elem(map, key, init, 0);
+	if (err)
+		return (void *) 0;
+
+	return bpf_map_lookup_elem(map, key);
+}
+
+SEC("tracepoint/raw_syscalls/sys_enter")
+int tracepoint__raw_syscalls__sys_enter(struct trace_event_raw_sys_enter *args)
+{
+	u64 id;
+	u64 ts;
+
+	/* Currently there is no "skel" way to skip load/attach
+	 * if !measure_latency, so check it here */
+	if (!measure_latency)
+		return 0;
+
+	id = bpf_get_current_pid_tgid();
+	if (filter_pid && id >> 32 != filter_pid)
+		return 0;
+
+	ts = bpf_ktime_get_ns();
+	bpf_map_update_elem(&start, &id, &ts, 0);
+	return 0;
+}
+
+SEC("tracepoint/raw_syscalls/sys_exit")
```

Comment:
Am I right that in the case of raw tracepoints we are getting an array of `u64`s pointing to arguments of `TP_PROTO`? In this case prototype for `sys_exit` tracepoint is `TP_PROTO(struct pt_regs *regs, long ret)` and doesn't include the `id` (syscall number). In kernel this is easily obtained as `syscall_get_nr(current, regs)`, but I am not sure how to do this in the BPF case (as it is arch-dependent).

Reply:
Yeah, it's very strange, but sys_exit does not have syscall id provided... You are right that it seems to be arch-dependent. You can get syscall id using pt_regs->orig_ax, but that would break on non-x86. I guess just stick to tracepoint then.